Kafka/src/main/java/io/deephaven/kafka/publish/PublishToKafka.java

```diff
@@ -0,0 +1,138 @@
+                final WritableObjectChunk<Future<RecordMetadata>, Attributes.Values> sendFutures =
+                        WritableObjectChunk.makeWritableChunk(chunkSize)) {
+            while (okit.hasMore()) {
+                final OrderedKeys chunkOk = okit.getNextOrderedKeysWithLength(chunkSize);
+
+                final ObjectChunk<K, Attributes.Values> keyChunk;
+                if (keyContext != null) {
+                    keyChunk = keySerializer.handleChunk(keyContext, chunkOk, usePrevious);
+                } else {
+                    keyChunk = null;
+                }
+
+                final ObjectChunk<V, Attributes.Values> valueChunk;
+                if (valueContext != null) {
+                    valueChunk = valueSerializer.handleChunk(valueContext, chunkOk, usePrevious);
+                } else {
+                    valueChunk = null;
+                }
+
+                sendFutures.setSize(0);
+                for (int ii = 0; ii < chunkOk.intSize(); ++ii) {
+                    final ProducerRecord<K, V> record = new ProducerRecord<>(topic,
+                            keyChunk != null ? keyChunk.get(ii) : null, valueChunk != null ? valueChunk.get(ii) : null);
+                    Future<RecordMetadata> x = producer.send(record);
+                    sendFutures.add(x);
+                }
+
+                // TODO: this makes us jerks for doing this midstream instead of waiting all the way until the end, but
+                // we shouldn't just forget about them either; maybe we need to keep a few chunks of futures around, or
+                // even we should be willing to have the whole thing be futured
+                for (int ii = 0; ii < chunkOk.intSize(); ++ii) {
+                    try {
+                        sendFutures.get(ii).get();
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException("Interrupted while waiting for KafkaProducer send result",
+                                e.getCause());
+                    } catch (ExecutionException e) {
+                        throw new RuntimeException("Failure while sending to KafkaProducer", e.getCause());
+                    }
+                }
+            }
+        }
+    }
+
+    private class PublishListener extends InstrumentedShiftAwareListenerAdapter {
+        private final boolean isStream;
+
+        private PublishListener() {
+            super("PublishToKafka", (DynamicTable) table, false);
+            this.isStream = StreamTableTools.isStream(table);
+        }
+
+        @Override
+        public void onUpdate(Update upstream) {
+            Assert.assertion(upstream.shifted.empty(), "upstream.shifted.empty()");
+
+            if (isStream) {
+                Assert.assertion(upstream.modified.empty(), "upstream.modified.empty()");
+                publishMessages(upstream.added, false, true);
```

Comment:
This seems to assume value is always going to be generated?  What about publishing just in Key and ignoring Value?